compiler/rustc_query_system/src/query/job.rs

```diff
@@ -29,73 +30,77 @@ pub struct QueryInfo<Q> {
     pub query: Q,
 }
 
-pub type QueryMap<Q> = FxHashMap<QueryJobId, QueryJobInfo<Q>>;
+pub(crate) type QueryMap<D, Q> = FxHashMap<QueryJobId<D>, QueryJobInfo<D, Q>>;
 
 /// A value uniquely identifiying an active query job within a shard in the query cache.
 #[derive(Copy, Clone, Eq, PartialEq, Hash)]
 pub struct QueryShardJobId(pub NonZeroU32);
 
 /// A value uniquely identifiying an active query job.
 #[derive(Copy, Clone, Eq, PartialEq, Hash)]
-pub struct QueryJobId {
+pub struct QueryJobId<D> {
     /// Which job within a shard is this
     pub job: QueryShardJobId,
 
     /// In which shard is this job
     pub shard: u16,
+
+    /// What kind of query this job is.
+    //FIXME: This seems to be unused, but removing it changes query cycle errors.
+    pub kind: D,
```

Comment:
What if its used by the `Eq` impl?

Reply:
Ohhhhhh, that must be it! Thanks!
I just didn't understand what was going on.